### DISCUSSION TITLE
Error accessing device after a few hours working

### DISCUSSION THREAD

**Discussion by vjbutt** [jasonacox/tinytuya#443] [Q&A]
I am getting the status from a WiFi Dual Meter every 10s using tinytuya.  Everything works fine for a few hours.  Then I start getting the following error:

Unexpected error: <class 'KeyError'>

I can still read the meter data using the Smart Life app whilst I am getting this error.  However, in order to get the status working again using tinytuya, I have to power cycle the WiFi Dual Meter.  Any idea what could be wrong?

**Comment by jasonacox** [comment]
Hi @vjbutt can you detail or link about the Tuya device (WiFi Dual Meter)?  

I have had a few older devices that would stop responding to the local network requests but would continue to work with the cloud (Smart Life). Power cycling them was the only that that would fix them. 

It would be interesting to see if you could run a monitoring loop with a persistent connect to see if that helps.  See https://github.com/jasonacox/tinytuya/blob/master/examples/monitor.py as an example.

**Comment by vjbutt** [comment]
@jasonacox thanks for your suggestions, I will try that plus a few other things.  It seems that I have bought one of these older devices :(.

The error that I reported above is wrong (that was actually due to a missing DP ID in one of the returned data).  The actual error that I am getting when the WiFi Dual Meter gives up on the local connection is:

{'Error': 'Check device key or version', 'Err': '914', 'Payload': None}

The device key is definitely correct, because when I power cycle the device it works.  I will next check the version.  I am currently setting this to 3.4, as is reported by the device, but maybe I should set this to 3.3.

**Comment by vjbutt** [comment]
@jasonacox I have tried setting the version 3.3, but tinytuya does not even manage to get the status.  So the correct value to set is actually 3.4.

I still need to find some time to check with a monitoring loop over a persistent connection.

BTW the device is this one - [https://www.amazon.co.uk/Energy-Electricity-Monitor-Control-Management/dp/B0CDXF3J1K/ref=cm_cr_arp_d_product_top?ie=UTF8](url)

Haven't managed to find a link to the manufacturer though :-(

**Reply by uzlonewolf** [reply]
I have some v3.4 devices which do this, and my suspicion is memory corruption within the device is causing it to "forget" what the correct local key is.  How often it happens is pretty random; one day it may do it 3 times that day, and then it will be fine for a month.  I haven't found a solution other than rebooting it.

**Comment by vjbutt** [comment]
@uzlonewolf that was exactly what I was thinking, that somehow there must be some memory corruption.  However, today I was surprised to note that the device started to respond to status updates again after about 1 hour of giving {'Error': 'Check device key or version', 'Err': '914', 'Payload': None} (without necessitating a power cycle).  So maybe it is not memory corruption.

Furthermore, just before it started to respond correctly again to status queries, it returned the following status:

{'protocol': 4, 't': 20, 'data': {'dps': {'2': 1887}}, 'dps': {'2': 1887}}

It seems to me that the device is somehow switching to protocol 4.  Why should it do this is beyond me.  The problem is that tinytuya does not support version 4, so maybe those error messages were being generated because the device switched to version 4.  Note that these protocol 4 messages are being received also in between correctly received status updates.  They seem to be returned quite at random.

One other thing to note is that the time stamp ('t':20) returned in the above status seems to indicate that the device had rebooted.  From previous protocol 4 status messages, it is clear that this is a Unix timestamp.  So a timestamp of 20 seems to indicate that not even the real-time clock on board of the device had yet been updated before this status was returned. This may indicate a more serious case of memory corruption.

Will keep monitoring.

**Reply by uzlonewolf** [reply]
The 'protocol' in the payload does not do anything as far as I can tell and is only present in packets which contain a "protocol version" header directly before the payload.  Devices set it to 4 while the SmartLife app sets it to 5.

**Comment by vjbutt** [comment]
@uzlonewolf, indeed that is now my impression as well.  The presence of these protocol 4 packets increases drastically when I've set set_socketPersistent(True).  They basically contain a very small subset of the DPS data that eventually follows in a full data payload.  

Interestingly, setting a persistent connection also seems to have solved my problem in that the device is still alive and supplying "valid" data after more than 24 hours of continuous operation.  However, it has now created a new problem.  The status command is returning repeated payloads.  I still see the measurements being updated on the Smart Life app, but not in tinytuya.  As a consequence the returned data is practically useless.  I have tried to force the update of the DPS using the function updatedps() but this does not seem to resolve the issue. 

Why would a persistent connection cause repeated payloads and not return the actual data being measured?  Note that I am only reading the status every 10s, which is compatible with the value set in the report rate control.

**Comment by Dronec** [comment]
Got the same situation with this device:
https://www.ebay.com.au/itm/355760971428
Querying it every 5 min, after approximately 12 hours it stops responding with
"Check device key or version"

**Comment by dwmw2** [comment]
I have a similar problem with a [Ciarra cooker hood](https://www.amazon.co.uk/dp/B0C6TJXJX1). I was connected to it and it was happily reporting status, and suddenly it started returning the same error. I was running this:
```
#!/bin/python3

import tinytuya
d = tinytuya.OutletDevice(dev_id='xxx', address='172.31.31.209', local_key='xxx', version=3.4)


data = d.status()
print('Received Payload: %r' % data)
while (True):
      data = d.receive()
      print('Received Payload: %r' % data)
      if not data:
        print(" > Send Heartbeat Ping < ")
        d.heartbeat()
      print(" > Request Update for DPS indexes 18, 19 and 20 < ")
      result = d.updatedps([4, 5, 6, 18, 19, 20])
      data = d.status()
      print('Received Payload: %r' % data)

```
.. and it just stopped:
```
$ python3 hood.py 
Received Payload: {'dps': {'1': False, '2': '1', '6': False, '12': '0', '15': 66, '104': 'Level_1'}}
Received Payload: None
 > Send Heartbeat Ping < 
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'dps': {'1': False, '2': '1', '6': False, '12': '0', '15': 66, '104': 'Level_1'}}
Received Payload: None
 > Send Heartbeat Ping < 
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'dps': {'1': False, '2': '1', '6': False, '12': '0', '15': 66, '104': 'Level_1'}}
Received Payload: None
 > Send Heartbeat Ping < 
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'dps': {'1': False, '2': '1', '6': False, '12': '0', '15': 66, '104': 'Level_1'}}
Received Payload: None
 > Send Heartbeat Ping < 
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
 > Request Update for DPS indexes 18, 19 and 20 < 
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
Received Payload: {'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
 > Request Update for DPS indexes 18, 19 and 20 < 
```
Now it reports the same error every time I try to connect. The Android app was working, and a power cycle seems to have fixed it for tinytuya too.

**Reply by Dronec** [reply]
Android app is working via the cloud, which doesn't seem to be affected by this issue. I'm starting thinking that emulating the cloud locally might be the way forward https://github.com/kueblc/mocktuyacloud

**Reply by jasonacox** [reply]
@dwmw2  - If you do not send `d.updatedps([4, 5, 6, 18, 19, 20])` does it still occur? That has always been a bit of a hack to try to get these devices to update those DPS values. I wonder if it is tripping some vulnerability in the firmware (e.g. memory leak) that eventually causes it to become unstable.

@Dronec - Yes, valid point. These devices were not designed for local access as we have seen again and again. They are designed for the cloud so that tends to be the most polished and reliable aspect. I think that project you reference has an interesting idea, but doesn't look like it has been touched in 5 years.

**Reply by Dronec** [reply]
@jasonacox I've tried that without sending updates, same result.

I'm going to play with that project over the weekend, just need to wait for a new device

**Reply by jasonacox** [reply]
I guess one experiment would be to remove heartbeat to see if that works. There is a high chance that it would hit heartbeat timeout and disconnect the TCP session.

And another experiment would be to stop sending anything at all and test to see if it is just a "running time" related issue vs some bug in their local API handler that traffic aggrevates. You are still having to test it every x minutes/hours, but it would at least tel lus that this is API usage related vs a firmware fundamental flaw.

Another experiment would be to write a timer in your loop to fully disconnect (tear down TCP link) and reconnect. Theory would be that whatever resource leakage that is causing the device to become unresponsive could be averted if it still does a cleanup after a connection close. Of course that also means that once it becomes unresponsive, it is no longer able to recover itself so picking the right timer for the reset before that threshold would be important.

**Comment by oeboema** [comment]
Same here, with this heater: https://eurom.nl/en/product/permanent-heating/bathroom-heaters-permanent/sani-bathroom-radiator-1000-wifi-white/

Stops responding after a while (several hours), and complains about the key not being valid. 2 Smartplugs that I'm polling work fine. A power cycle for the heater does resolve the situation (for a while).

**Reply by jasonacox** [reply]
Hi @oeboema can you share what you are using to access the device (script or just scanning)?

**Comment by AlphaZulu666** [comment]
same problem with a Wifi Power meter, it worked for a month, then started like this.... the bad is that I couldn't find a way to powercycle it, even from the app, and that's a mess, because it is needed a physical powercycle.
Isn't there a way to do the same calls the app do to get the data?

**Reply by Dronec** [reply]
The app is working via the cloud, so you can extract the data from the cloud. I'm in exactly the same situation: failed wifi circuit breaker, which I'm getting the data from via the cloud.

**Reply by jasonacox** [reply]
Request: Please share a link to the device (e.g. amazon, aliexpress) if you can. I have found no way to remote power cycle any Tuya device. If anyone ever discovers this, please let us know.

**Reply by Dronec** [reply]
> Request: Please share a link to the device (e.g. amazon, aliexpress) if you can. I have found no way to remote power cycle any Tuya device. If anyone ever discovers this, please let us know.

This is mine: https://www.ebay.com.au/itm/355760971428

**Comment by elljay234** [comment]
Hi, similar issue with a unit described as "Tuya Smart 2 Way WiFi Energy Meter Bidirection 2 Channel with Clamp" I bought from ebay in 2023.  The listing has been removed so I can't post a link.

It normally works for weeks or months then seems to need a power cycle to get it going again.  Yesterday and today it's done it several times!

I get a mixture of these errors:
{'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
{'Error': 'Network Error: Unable to Connect', 'Err': '901', 'Payload': None}
{'Error': 'Network Error: Device Unreachable', 'Err': '905', 'Payload': None}

I can't see any model number on the case of the device - only the front is printed:

![image](https://github.com/user-attachments/assets/b5953e13-58eb-4a1e-a05f-653ccbbe99af)

Shout if you want more details / want me to run some debug code...

**Comment by Kowa92** [comment]
Same issue with [EARU Eelectric Wi-Fi Two-way Energy Meter](https://www.aliexpress.com/item/1005005598295699.html?spm=a2g0o.order_list.order_list_main.26.63f11802RE2xph). The first two weeks they worked well.. Now if they work a few hours without giving me the 914 error, it's already an achievement. The only way to get them to work is to cut off the power and then turn it back on (always praying that they will work at least for a while).

- Has anyone by any chance found a solution to this problem?

Post the debug output of my code:

```
DEBUG:TinyTuya [1.16.1]

DEBUG:Python 3.9.2 (default, Dec  1 2024, 12:12:57) 
[GCC 10.2.1 20210110] on linux
DEBUG:Using pyca/cryptography 3.3.2 for crypto, GCM is supported
DEBUG:status() entry (dev_type is default)
DEBUG:final payload_dict for 'bf8b89b198c36175bdhf3x' ('v3.4'/'default'): {1: {'command': {'gwId': '', 'devId': '', 'uid': '', 't': ''}}, 7: {'command': {'protocol': 5, 't': 'int', 'data': {}}, 'command_override': 13}, 8: {'command': {'gwId': '', 'devId': ''}}, 9: {'command': {'gwId': '', 'devId': ''}}, 10: {'command': {}, 'command_override': 16}, 13: {'command': {'protocol': 5, 't': 'int', 'data': {}}}, 16: {'command': {}}, 18: {'command': {'dpId': [18, 19, 20]}}, 64: {'command': {'reqType': '', 'data': {}}}}
DEBUG:building command 10 payload=b'{}'
DEBUG:sending payload quick
DEBUG:final payload: b'0123456789abcdef'
DEBUG:payload encrypted=b'000055aa000000010000000300000044fb87b1927b9a5e829bce171f540886e0393d8f2cb924c51ec39b1d8b918945aab01a9ba075e10fddf56ec7b77bfca6c1bfe9266c54c044c407e5407d873ec8b40000aa55'
DEBUG:session key negotiation failed on step 1
DEBUG:ERROR Check device key or version - 914 - payload: null
DEBUG:status() received data={'Error': 'Check device key or version', 'Err': '914', 'Payload': None}
```

Thanks

**Reply by jasonacox** [reply]
Does the device still respond to SmartLife app?

**Reply by Kowa92** [reply]
> Does the device still respond to SmartLife app?

Yes, the device still respond to SmartLife.

**Reply by jasonacox** [reply]
Thanks, @Kowa92 . I wonder if this is due to a polling problem. If you reduce or disable the polling of the device (no TinyTuya or localtuya calls) does it last longer?

**Reply by Dronec** [reply]
> Thanks, @Kowa92 . I wonder if this is due to a polling problem. If you reduce or disable the polling of the device (no TinyTuya or localtuya calls) does it last longer?

I can confirm that if you call it only occasionally, it lasts much longer.

**Reply by jasonacox** [reply]
If someone has time to do packet traces, especially comparing TinyTuiya and  SmartLife app use (local mode), it would be interesting to see if there is any indication of additional handshakes we are missing (e.g. causing the device ro recover resources). I suspect not and that repeating SmartLife app local traffic use may eventually cause the same thing (indicating bad code on the device for local API but good for cloud API). But maybe some comparison would identify traffic patterns we could adjust. Without seeing the actual code the device is running, we are fishing around in the dark on why it fails (doesn't recover) from basic local API use.

**Reply by Dronec** [reply]
> If someone has time to do packet traces, especially comparing TinyTuiya and SmartLife app use (local mode), it would be interesting to see if there is any indication of additional handshakes we are missing (e.g. causing the device ro recover resources). I suspect not and that repeating SmartLife app local traffic use may eventually cause the same thing (indicating bad code on the device for local API but good for cloud API). But maybe some comparison would identify traffic patterns we could adjust. Without seeing the actual code the device is running, we are fishing around in the dark on why it fails (doesn't recover) from basic local API use.

Jason, can you please tell me more about SmartLife "local" mode? I did packet tracing it once (not too thoroughly though) and didn't notice any local communications apart of the initial registration, the rest was going through the cloud

**Reply by uzlonewolf** [reply]
My problematic v3.4 devices usually last weeks if not months before dropping offline, however I make a single persistent connection to them that lasts that entire time.  When they do drop offline the SmartLife app no longer works in local mode either.  I've thought about trying to packet capture them but with my months-long connection it just hasn't been practical.  I'll need to see if I can scare up another problematic device that I can put on my test network.

**Comment by elljay234** [comment]
Not sure if it helps, but just looking in the Tuya cloud Device Log I can match the timing of a locally reported set of "Network Error: Unable to Connect', 'Err': '901'" errors with the device reporting to Tuya that it was offline due to a "keepalive_timeout".  The online Tuya log only seems to go back seven days - my last Err 914 was almost two weeks ago!  I'll have a look at the log next time I get a 914...

(The online log is here: https://platform.tuya.com/cloud/ - "Open Project", "Devices", "Debug Device", "Device Logs")

**Reply by elljay234** [reply]
Had a 914 error overnight.  The log reports "Event: Device semaphore"; "Event Details: signal intensity:-60||Remaining memory:32240||UTC:1742868113||DST:1742868113", shortly followed by "Device online".  The device carried on working and didn't need a power cycle.  Smart Life says WiFi signal strength -50 to -70 is "Good", so maybe it was low on memory...?

**Reply by jasonacox** [reply]
Interesting... - "Device semaphore": This may refer to a synchronization mechanism used to manage access to shared resources. A semaphore is a variable that controls the access to a shared resource by multiple processes or threads. In this case, the device semaphore is likely being used to manage access to the device  state? I can only speculate, but I wonder if this means some thread locking mechanism failed and maybe 914 just restarts.

**Reply by elljay234** [reply]
Had a look at the log in more detail...  It only allows you to filter by one event type at a time so I copied and pasted everything except the "EVENT_TYPE_REPORT" which are the measured data into Excel and sorted by date.

First observations are:
- It reports "Online" every 12 hours.  Network issues restart that routine.
- Remaining memory remains constant between "Online" events. (Except one occasion 24/3 06:33!)
- It reports "Device restart - power off reboot" when I've not power cycled the device.
- The one 914 Tinytuya error (which didn't need a power cycle) is one second after the device reported an "online" event.

Have attached the Excel file and a screenshot of the file:
![TuyaDeviceLog-March](https://github.com/user-attachments/assets/eb27b7aa-316f-470c-bd34-22f6da0f5a10)
[TuyaDeviceLog-March.xlsx](https://github.com/user-attachments/files/19464167/TuyaDeviceLog-March.xlsx)

**Comment by Dronec** [comment]
Got this error after 2 days of querying thing:

https://www.aliexpress.com/item/1005004153509348.html

Interesting thing is that I was keeping the connection open to avoid memory leaks and stuff.

{"Error": "Check device key or version", "Err": "914", "Payload": null}

**Comment by Dronec** [comment]
However, after that something strange and miraculous started happening. I've noticed that when I open the mobile app (Smart Life) the device starts responding again (and stops when I close it).

![image](https://github.com/user-attachments/assets/29e8dea6-bcf0-49cd-98f8-8bf8f0a4f198)

**Comment by sueastside** [comment]
@Dronec Can you share a bit more about your setup? I cannot replicate this behavior with the Smart Life app....

I know updating the firmware restarts the device and it starts working again, any way we can trigger this reboot remotely as well?

**Reply by Dronec** [reply]
I'm starting thinking that this might strongly depend on a brand of the device or even a specific device. I have 4 circuit breakers bought on AliExpress under the brand TONGOU and they all behave differently.
My setup is basically a piece of code from tinytuya examples (in repository, see threading) which writes the responses into a database day and night. One of the devices works totally as expected, other three randomly throw error 914, go offline (so even Smart life app can't see them), and then back online like nothing happened. Happy to share more details if needed

**Comment by uzlonewolf** [comment]
A few months ago I blocked multicast/broadcast from my IoT SSID (on the Unifi controller it's Settings -> Wireless Networks -> Edit -> "Block LAN to WLAN Multicast and Broadcast Data" checked).  Ever since then, not one v3.4 device has dropped offline, though I am still losing 1 or 2 v3.3 devices per week (these devices disconnect from the network completely and cannot be controlled by the cloud or the app).

Note however that blocking broadcasts on the wireless network means an app running on a phone connected wirelessly can no longer find them; to get around this you can run https://github.com/jasonacox/tinytuya/blob/master/tools/broadcast-relay.py on any computer connected via wire and it will relay the discovery packets to the phone app.

**Reply by marcsowen** [reply]
Very interesting. But I don't quite get it. You mean, broadcasts/multicasts TO your IoT network are being blocked? So your Tuya devices don't receive anything except unicast traffic from tinytuya? Or the other way around? Broadcasts transmitted by your Tuya devices are not being forwarded to other parts of your LAN? But how would this help your Tuya device? Would the Tuya device even notice that UDP broadcasts are being discarded?

Or is this somehow also blocking connections from your Tuya device to the Tuya cloud?

**Reply by uzlonewolf** [reply]
> You mean, broadcasts/multicasts TO your IoT network are being blocked? So your Tuya devices don't receive anything except unicast traffic from tinytuya?

Yes, exactly.  The only thing the devices see are unicast packets for them (both from TinyTuya and the Tuya cloud).

I don't know why it works, my assumption is the memory corruption is happening in the broadcast packet handler.  Devices listen for broadcasts from neighbors so they know about other devices on the network they can create scenes with.

**Reply by marcsowen** [reply]
> Yes, exactly. The only thing the devices see are unicast packets for them (both from TinyTuya and the Tuya cloud).
> 
> I don't know why it works, my assumption is the memory corruption is happening in the broadcast packet handler. Devices listen for broadcasts from neighbors so they know about other devices on the network they can create scenes with.

Strange. I just have one single Tuya device in my network - so no other Tuya devices are broadcasting as neighbors. I still get the error though. I helped myself by adding a simple proxy which keeps the connection to my device open in order to always get the latest measurements: https://github.com/marcsowen/tuyaproxy

Maybe some other broadcasts/multicasts in my network cause the error.

**Reply by Adrian-Rosoga** [reply]
> A few months ago I blocked multicast/broadcast from my IoT SSID (on the Unifi controller it's Settings -> Wireless Networks -> Edit -> "Block LAN to WLAN Multicast and Broadcast Data" checked). Ever since then, not one v3.4 device has dropped offline, though I am still losing 1 or 2 v3.3 devices per week (these devices disconnect from the network completely and cannot be controlled by the cloud or the app).
> 

Hi @uzlonewolf, I could try going that path. My Zyxel router from my ISP (Hyperoptic) does not have such option (blocking multicast/broadcast).

Would an access point like this one below do the job? I would connect the heaters to the APs Wifi.

https://www.amazon.co.uk/Ubiquiti-Unifi-U6-without-Injector/dp/B0BYL7G7SZ

I looked up Unifi APs on Amazon, so I could return the AP if it doesn't solve the problem.

**Reply by uzlonewolf** [reply]
That should fix the v3.4 devices, but I have yet to find anything that keeps the v3.3 devices from dropping offline completely.

As a side note, keep in mind that UniFi devices do not have built-in admin interfaces and you are going to need to install the controller on a computer somewhere to configure them.  (The controller does not need to remain running, it only needs to run when making config changes or if you want logging.)

**Comment by Adrian-Rosoga** [comment]
Adding to this thread some extra info. Not being able to locally control the devices is a showstopper for my use case.

Err 914 with 2 (two!) out of 4 Ecostrad iQ Plus electric heaters (https://ecostrad.com/ecostrad-iq-plus-electric-panel-heater/).
3 of them are 3.3, one is 3.4.

Summary of the situation now:
- 1 heater shows Offline in SmartLife app - this needs returning, local & cloud control failure - it's the 2nd heater to return
- 1 heater 3.4 shows Err 914 - control via SmartLife app is OK
- 1 heater 3.4 shows Err 901 - control via SmartLife app is OK
- 1 heater 3.3 shows Err 901 - control via SmartLife app is OK

I hoped that changing to explicit IP address from Auto would help - nope. Maybe changing from Auto to explicit IP address (fixed IP) causes 901 after a while?
@uzlonewolf suggested disabling multicast - my router does not support that, buying an extra AP is one idea for later.
@jasonacox - if you have any suggestion please let know.

Any chance to bump up this issue and aim for a fix? I am interested in the technicalities and eager to help.
Let me know.

Is Tuya providing a development board or has someone a board where to play with?

**Reply by uzlonewolf** [reply]
I'm a bit confused, you say 3 out of 4 are v3.3 but only list 1?

Also, how many of these work for a bit and then stop vs how many have never worked?

If a device is completely offline even in the app then I'd try power cycling it and, if that doesn't work, factory defaulting it and re-adding it.  I have a number of v3.3 smart bulbs that occasionally disconnect from the WiFi and need to be power cycled to get them back; I have yet to find anything which fixes this problem.

Err 901 means an exception was thrown when trying to establish a TCP connection or when trying to transmit data packets.  This is usually caused by either the wrong IP address (did the DHCP server give the device a new IP?) or the maximum number of simultaneous TCP connections was reached and the device is refusing new connections until an existing one is closed.  Disabling multicast/broadcast will not fix this problem either.

v3.4 devices that work for a bit and then start throwing 914 errors later are the only ones that can be fixed by disabling WiFi multicast/broadcast.

None of these problems can be fixed by us, they need the device manufacturer to push out a firmware update.  If you wanted to load your own firmware then you would be better off just flashing something like OpenBeken and ditching Tuya completely.  LAN control for Tuya devices is an afterthought and they don't put much effort into it.

Tuya has at least a dozen different modules, which one are you interested in?  I have a whole bin of random boards and probably have one.  I also have a decent collection of different firmwares.  I've decompiled a few but they make heavy use of dynamically-registered callbacks which makes the program flow next to impossible to follow.

**Reply by Adrian-Rosoga** [reply]
Oh, I wasn't clear then: 3 are 3.4 and one is 3.3.

I am encountering a mixture of 2 different problems with these new Ecostrad radiators (avoid them!):
1. Tuya issues: Local control failing with Err 914 after some time. Getting now Err 901 as well.
2. More serious issues: Control board failure: Radiators showing Offline in Tuya app - that is, failure of both LAN & cloud. At least the manual control still works.

Lesson learned: Do not buy smart radiators! Buy dumb good quality ones and make your own DIY remote control.

As a plus, one can use a temperature sensor away from the body of the radiator. This is what I am working on in parallel with making the best of these new and expensive radiators ;).

I am in London btw. So you played with the hw and sw a lot. I was pretty naive - I even thought I would reverse engineer the protocol using Wireshark, etc. Went only as far as seeing the radiators connecting to AWS in Germany, the rest was too complicated, but I was happy finding tinytuya. In a few hours I had cloud and local control working, hooray. After a day though got 914 (1 above) and after some time the more comprehensive failure (2 above). Still having fun and hope!

**Reply by uzlonewolf** [reply]
> More serious issues: Control board failure: Radiators showing Offline in Tuya app - that is, failure of both LAN & cloud. At least the manual control still works.

Have you tried shutting power off to it and turning it back on?  I have some bulbs that do this and power cycling them brings them back.

**Reply by Adrian-Rosoga** [reply]
I am going reboot them, of course. Just that the place where the heaters are located is away, I can go there only now and then. In the meantime I monitor via VNC to a Pi 3 and via SmartLife.
Because the place is remote I'd need a rock solid control solution.

**Reply by vjbutt** [reply]
The only reliable workaround that I found is to feed my Tuya device through a smart switch.  I then monitor the Tuya device through a Python script.  As soon as I start getting errors, I power cycle the Tuya device through the smart switch.  This will involve the purchase of an extra smart switch and some extra coding, but it works.

**Reply by uzlonewolf** [reply]
Sadly I don't think you're going to get a rock solid solution while using the stock firmware.  Some v3.3 devices like to drop off the WiFi completely, and a lot of (most?) v3.4 devices "forget" their local key from time to time.  v3.5 is still too new to know if they are going to have any issues or not.  If you need something rock solid then you're probably better off flashing everything with OpenBeken or similar.
